### ISSUE TITLE
Add derives on derived FooBuilder struct

### ISSUE THREAD

**Issue by colin-kiegel** [colin-kiegel/rust-derive-builder#43]
# use case

User wants to have something like `PartialEq` for `FooBuilder`, but can't derive it since `FooBuilder` is already derived by itself.

# possible API

We could add a struct annotation on `Foo`:
- `#[builder(derive(PartialEq, SomethingElse))]`

# example

```rust
#[derive(Builder)]
#[builder(derive(PartialEq, SomethingElse))]
struct Foo {
    foo: Vec<i32>
}
```

Would expand to:
```rust
#[derive(Default, Clone, PartialEq, SomethingElse)]
struct FooBuilder {
    foo: Option<Vec<i32>>
}
```

# discussion

This crate should be conservative and derive as little as possible for `FooBuilder` by default. Because additional derives on `FooBuilder` would add corresponding requirements on the fields of `Foo`.

Status Quo:
- `FooBuilder` impls `Default` and `Clone` (the latter is only needed for the immutable builder pattern)
- All fields on `Foo` already require to implement ...
  - `Clone`, if they have a setter on `FooBuilder`
  - `Default`, if they _don't_ have a setter on `FooBuilder` (.. well, at least with '#15: hidden fields' coming, and this might change with '#3: CustomDefault')

**Comment by colin-kiegel** [comment]
PartialEq on `FooBuilder` might be useful for some unit tests.

**Comment by TedDriggs** [comment]
Enabling a builder to derive serde's Deserialize would be useful for the case of wanting to read an object off the wire (or from a file), applying some overrides, then returning a final constructed object.

**Comment by colin-kiegel** [comment]
Interesting use case.

I see one complication: serde has a lot of field attributes. It would look very ugly if we'd use some pass-through syntax for those field attributes.

However I think that use case can be solved differently. We could derive a second build method on `FooBuilder`, which takes two arguments `fn build_with_defaults(&self, defaults: &Foo) -> Foo` or maybe a mutating variant `fn apply_to(&self, target: &mut Foo)`. Modulo bike shedding. You could then deserialize directly to `Foo` and then apply some incompletely initialized builder to it. Would something like this help you? Or could you describe in more detail, why you need a builder _and_ deserialization at the same time? Where do your overrides come from (what do they depend on)?

PS: I am always happy to hear about what people need and want to achieve. :-)

**Comment by TedDriggs** [comment]
I'm writing a command-line app which reads from a config file but allows those values to be overridden with command-line arguments. Serde is great for deserializing from the file, but now I have to apply the command-line overrides. At the moment, that means having mutators on the config struct. That _works_, but it means that the struct is more mutable than I'd like.

The `apply_to` approach would work well for my case, though I could imagine that causing problems with RAII in other scenarios.

**Comment by TedDriggs** [comment]
Now that I'm more familiar with the code of this crate, I wonder if there might be a solution like this:

```rust
/// Policies and connection information for a whitelist manager instance.
#[derive(Debug, Clone, PartialEq, Deserialize, Builder)]
pub struct Config {
    pub appliance: Appliance,
    pub rules: Vec<Rule>,

    #[serde(default, rename = "limit")]
    #[builder(default, preserve_attr = "serde")]
    pub manual_limit: Option<usize>,

    #[serde(default, rename = "track_usage")]
    #[builder(default, preserve_attr = "serde")]
    pub track_indices: bool,
}
```

The `preserve_attr` field would carry the serde attribute forward unmodified to the field on the `ConfigBuilder` struct. At least for my use-case, I think that would be perfect. I could then deserialize into `ConfigBuilder`, invoke setter methods for the command-line overrides, call `build()` and proceed.

(For the project where I encountered this, I'm actually considering a refactor that would remove the immediate need for this. But generally I think the idea of being able to forward attributes from the struct to the builder would be great for the "deserialize and override an eventually-immutable object" pattern.)

**Comment by colin-kiegel** [comment]
Hm, I'd like to _avoid_ passing field attributes to the builder. Because that would make everything very hard to reason about. What if someone wants to have an attribute only on a FooBuilder field, but not on the Foo field?

I'd be more open to these things
- `derive`s on the builder, but without attributes.
- `apply_to` method as mentioned above.
- derive_builder could skip the struct definition and leave that to the user. It would then only generate the setters. The user would need to exactly match the signature and error messages will most likely be not very helpful due to lacking span information, like "_some_ type was not as expected .. ok, which one?". But on the other hand it's really just wrapping everything in an Option.

There will probably always be special cases, where derive_builder won't fit for some reason.

**Comment by colin-kiegel** [comment]
released with v0.4.5 - thanks @TedDriggs :-)

Let's open a separate ticket to discuss extensions of this feature.
